### PR TITLE
Add redis cluster and sentinel support

### PR DESCRIPTION
**Pull request by erdody**
Leveraging radix support for redis cluster/sentinel modes.

### CODE REVIEW

**Comment by erdody** [comment]
@mattklein123 new version of https://github.com/envoyproxy/ratelimit/pull/177. We finished with the integration test and a couple other things we found. Let us know if there's anything else we need to add. Thanks.

**Comment by mattklein123** [comment]
Thanks! @ysawa0 do you mind taking an initial pass?

**Comment by ysawa0** [comment]
@mattklein123 Can do. Taking a look.

File: test/integration/integration_test.go [hunk 1]
```diff
@@ -68,18 +68,45 @@ func TestBasicAuthConfig(t *testing.T) {
...
+func TestBasicReloadConfigWithRedisCluster(t *testing.T) {
+	t.Run("BasicWithoutWatchRoot", testBasicConfigWithoutWatchRootWithRedisCluster("8096", "false", "0"))
+	t.Run("ReloadWithoutWatchRoot", testBasicConfigReloadWithRedisCluster("8098", "false", "0", "false"))
+}
+
+func TestBasicReloadConfigWithRedisSentinel(t *testing.T) {
+	t.Run("BasicWithoutWatchRoot", testBasicConfigWithoutWatchRootWithRedisSentinel("8296", "false", "0"))
+	t.Run("ReloadWithoutWatchRoot", testBasicConfigReloadWithRedisSentinel("8298", "false", "0", "false"))
+}
+
```
Comment (tinder-yukisawa) [inline test/integration/integration_test.go:99]: I'm not sure if we need these additional tests as the config reload logic and redis logic should be in separate code paths and applying different settings for `RUNTIME_WATCH_ROOT` and `REDIS_xxx` shouldn't affect one another. I guess it doesn't hurt?

Comment (ysawa0) [inline test/integration/integration_test.go:99]: I'm not entirely sure if we need these additional tests as the config reload logic and redis logic should be in separate code paths and applying different settings for `RUNTIME_WATCH_ROOT` and `REDIS_xxx` shouldn't affect one another. I guess it doesn't hurt?

File: README.md [hunk 2]
```diff
@@ -501,6 +502,20 @@ As well Ratelimit supports TLS connections and authentication. These can be conf
 1. `REDIS_TLS` & `REDIS_PERSECOND_TLS`: set to `"true"` to enable a TLS connection for the specific connection type.
 1. `REDIS_AUTH` & `REDIS_PERSECOND_AUTH`: set to `"password"` to enable authentication to the redis host.
 
+## Redis type
+
+Ratelimit supports different types of redis deployments:
+
+1. Single instance (default): Talk to a single instance of redis, or a redis proxy (e.g. https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_protocols/redis)
+1. Sentinel: Talk to a redis deployment with sentinel instances (see https://redis.io/topics/sentinel)
+1. Cluster: Talk to a redis in cluster mode (see https://redis.io/topics/cluster-spec)
```
Comment (ysawa0) [inline README.md:511]: Nitpick -- should be 2 and 3 although Markdown renders them correctly as 2, 3

**Comment by ysawa0** [comment]
Everything else looks great!

**Comment by mattklein123** [comment]
Awesome, LGTM modulo the one comment above.

File: test/integration/integration_test.go (hunk 1 above)
Comment (erdody) [reply test/integration/integration_test.go]: Sounds good, removing.

**Comment by erdody** [comment]
Thanks @ysawa0 @mattklein123. I removed the test and made a few minor cleanups. Please take a look when you have a chance.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**